clients/da-vinci-client/src/main/java/com/linkedin/davinci/notifier/PushStatusNotifier.java

```diff
@@ -118,17 +119,51 @@ public void dataRecoveryCompleted(String kafkaTopic, int partitionId, long offse
   public void startOfIncrementalPushReceived(String topic, int partitionId, long offset, String message) {
     offLinePushAccessor
         .updateReplicaStatus(topic, partitionId, instanceId, START_OF_INCREMENTAL_PUSH_RECEIVED, offset, message);
-    updateIncrementalPushStatus(topic, message, partitionId, START_OF_INCREMENTAL_PUSH_RECEIVED);
+    updateIncrementalPushStatusToPushStatusStore(topic, message, partitionId, START_OF_INCREMENTAL_PUSH_RECEIVED);
   }
 
   @Override
   public void endOfIncrementalPushReceived(String topic, int partitionId, long offset, String message) {
     offLinePushAccessor
         .updateReplicaStatus(topic, partitionId, instanceId, END_OF_INCREMENTAL_PUSH_RECEIVED, offset, message);
-    updateIncrementalPushStatus(topic, message, partitionId, END_OF_INCREMENTAL_PUSH_RECEIVED);
+    updateIncrementalPushStatusToPushStatusStore(topic, message, partitionId, END_OF_INCREMENTAL_PUSH_RECEIVED);
   }
 
-  private void updateIncrementalPushStatus(
+  @Override
+  public void batchEndOfIncrementalPushReceived(
+      String topic,
+      int partitionId,
+      long offset,
+      List<String> pendingReportIncPushVersionList) {
+    /**
+     * We will perform filtering on batch inc push to report, as by original design we will only keep latest N inc push
+     * entries.
+     * For system store dual write, this is not perfect but good enough for now to keep a low volume of updates without
+     * refactor the key schema.
+     */
+    List<String> filteredIncPushVersionList = pendingReportIncPushVersionList;
+    if (pendingReportIncPushVersionList.size() > MAX_INCREMENTAL_PUSH_ENTRY_NUM) {
+      filteredIncPushVersionList = pendingReportIncPushVersionList.subList(
+          pendingReportIncPushVersionList.size() - MAX_INCREMENTAL_PUSH_ENTRY_NUM,
+          pendingReportIncPushVersionList.size());
+    }
+    offLinePushAccessor
+        .batchUpdateReplicaIncPushStatus(topic, partitionId, instanceId, offset, filteredIncPushVersionList);
+    for (String incPushVersion: filteredIncPushVersionList) {
+      updateIncrementalPushStatusToPushStatusStore(
```

Comment:
Do we have to report SOIP status to system store?

Reply:
In theory, no, but let me double check whether there is any implication with SOIP during VPJ polling.